ZOOKEEPER-2628: Fix findbug warnings.

hanm opened the pull request:
This PR fixed 19 find bug warnings and disabled one find bug warning:

Malicious code vulnerability Warnings
org.apache.zookeeper.ZooDefs$Ids.OPEN_ACL_UNSAFE is a mutable collection
Bug type MS_MUTABLE_COLLECTION (click for details) 
In class org.apache.zookeeper.ZooDefs$Ids
Field org.apache.zookeeper.ZooDefs$Ids.OPEN_ACL_UNSAFE
At ZooDefs.java:[line 116]

We will use ZOOKEEPER-1362 for fixing this specific warning, which might require change to ZooDefs interface.

hanm commented on src/java/main/org/apache/jute/compiler/CGenerator.java:
```diff
@@ -61,70 +61,88 @@ void genCode() throws IOException {
+                    }
+                }
+            }
+            if (c != null) {
```
It's sad that we have to close c twice (one here, one in the finally block earlier.). Findbugs is not happy if either of them is missing. It could be that the flow sensitive engine of findbug need an improvement.

breed replied:
i think try with resources would clean this up. we are on java7 now right?

hanm replied:
Yeah I thought about using try / with  - for some reasons I had the impression that we also need to support JDK 1.6 (that's probably only for 3.4.x). I'll update with try / with.

breed commented on src/java/main/org/apache/zookeeper/cli/DeleteCommand.java:
```diff
@@ -59,11 +59,6 @@ private void retainCompatibility(String[] cmdArgs) throws CliParseException {
         // delete path [version]
         if (args.length > 2) {
             // rewrite to option
```
you can remove the comment too :)

breed commented on src/java/main/org/apache/zookeeper/version/util/VerGen.java:
```diff
@@ -93,6 +93,14 @@ public static void generateFile(File outputDir, Version version, int rev, String
         } catch (IOException e) {
             System.out.println("Unable to generate version.Info file: "
                     + e.getMessage());
+            if (w != null) {
```
wow this is a weird one... we have it in the finally... try with resources would make this a bit nicer too...

hanm replied:
Findbug complains here because we use System.exit in catch block - in that case it is not guaranteed that the finally block will be executed.

hanm commented:
Updated PR to address review comments from @breed.

eribeiro commented on src/java/main/org/apache/jute/compiler/JRecord.java:
```diff
@@ -141,109 +141,128 @@ public String genCsharpWriteWrapper(String fname, String tag) {
+        } catch (IOException e) {
+            throw e;
+        } finally {
+            if (h != null) {
```
Just speculating here: couldn't we write this snippet as:

```
private IOException maybeClose(FileWriter file) {
       IOException t = null;
       if (file != null) {
             try {
                     file.close();
             } catch (IOException ex) {
                   t = ex;
             }          
     }
    return t;
}

(...)

} finally {

    IOException e1 = maybeClose(h);
    IOException e2 = maybeClose(c);

    if (e1 != null) {
        throw e1;
    }

   if (e2 != null) {
       throw e2;
   }
}


```

hanm replied:
updated with a more concise version that findbug is happy with.

hanm commented:
Updated PR to address review comments from @eribeiro.

lvfangmin commented on src/java/main/org/apache/zookeeper/server/DataTree.java:
```diff
@@ -1293,13 +1293,12 @@ public synchronized WatchesSummary getWatchesSummary() {
+            pwriter.print("0x" + Long.toHexString(entry.getKey()));
             pwriter.println(":");
-            HashSet<String> tmp = ephemerals.get(k);
+            HashSet<String> tmp = entry.getValue();
```
Usually, prefer to use the interface if we only use the methods defined in the interface, which makes it flexible (or less change needed) in case we want to use another Set implementation. 

There is a best practice defined in Effective Java 2nd Edition, Item 52: Refer to objects by their interfaces:
If appropriate interface types exist, then parameters, return values, and fields should all be declared using interface types. If you get into the habit of using interface types, your program will be much more flexible. It is entirely appropriate to refer to an object by a class if no appropriate interface exists.

eribeiro replied:
Ya @lvfangmin, one of my favourite books on Java, and was the reference to my comment above. Thanks pointing out. :smiley:

Agree with @hanm, better to create a separate issue for dealing with this legacy technical debt. :+1:

lvfangmin commented on src/java/main/org/apache/zookeeper/server/command/FourLetterCommands.java:
```diff
@@ -151,9 +153,13 @@
     public final static int telnetCloseCmd = 0xfff4fffd;
 
-    public final static HashMap<Integer, String> cmd2String =
+    final static HashMap<Integer, String> cmd2String =
```
Same here, we can also define this as Map to keep consistent with the others.

eribeiro replied:
We just need to make sure this doesn't break any client exposed public field. Looks like it is **not** this case (that is, we can make it a `Map`), but unfortunately I guess not all concrete declarations in ZK codebase can be changed to interfaces without breaking backward compatibility. It is not that we are not going to change it, but probably only at major release version.

fpj replied:
If we can remove this `public`, then I think we should. Also agree with the consistent `Map` declaration comment.

hanm commented:
Update: disable "Malicious code vulnerability Warnings" appertains to org.apache.zookeeper.ZooDefs$Ids.OPEN_ACL_UNSAFE, and we will use ZOOKEEPER-1362 for this work.

Can we get this merged? Would be good to have findbug clean again for pre-commit builds.

fpj commented on src/java/main/org/apache/zookeeper/cli/DeleteCommand.java:
```diff
@@ -56,14 +56,7 @@ public CliCommand parse(String[] cmdArgs) throws CliParseException {
-        // delete path [version]
         if (args.length > 2) {
-            // rewrite to option
-            String [] newCmd = new String[4];
```
Why are we removing this rewrite?

hanm replied:
These are dead code. newCmd was declared but not used so it's better to remove them.

fpj commented on src/java/test/config/findbugsExcludeFile.xml:
```diff
@@ -144,4 +144,10 @@
     <Bug pattern="DM_DEFAULT_ENCODING" />
   </Match>
 
+  <!-- Disable 'Malicious code vulnerability warnings' due to mutable collection types in interface.
```
We probably want to leave a note in ZOOKEEPER-1362 to remind ourselves to do it.

eribeiro replied:
Gonna say that! At least a TODO mark I guess...

hanm replied:
Comments added to ZOOKEEPER-1362.

fpj commented on src/java/main/org/apache/jute/compiler/JRecord.java:
```diff
@@ -576,174 +580,174 @@ public void genCsharpCode(File outputDirectory) throws IOException {
                 JField jf = i.next();
-                cs.write(jf.genCsharpCompareTo());
-                cs.write("    if (ret != 0) return ret;\n");
+                cs.write(jf.genCsharpEquals());
```
What's the reason for changing to `genCsharpEquals`? I suppose is is the same reason for changing to `genJavaEquals`...

hanm replied:
Similarly, there is no actual change here.

fpj commented:
@hanm I've left a few comments, but in general looks pretty good. I noticed that there are also unaddressed comments from @lvfangmin and @eribeiro . Could you take care of those so that we can check this in, please?

hanm commented on src/java/main/org/apache/jute/compiler/JRecord.java:
```diff
@@ -403,168 +410,165 @@ public void genJavaCode(File outputDirectory) throws IOException {
+                jj.write("    int ret = 0;\n");
+                for (Iterator<JField> i = mFields.iterator(); i.hasNext(); fIdx++) {
+                    JField jf = i.next();
+                    jj.write(jf.genJavaCompareTo());
```
This diff presented by git is not very accurate to reflect the state of the changes. There is no literal changes to these code blocks except indentations. We can compare the vanilla source file in master branch and in my working branch where the PR is created to double check:
JRecord.java in master: https://github.com/apache/zookeeper/blob/master/src/java/main/org/apache/jute/compiler/JRecord.java#L525
JRecord.java in PR branch:
https://github.com/hanm/zookeeper/blob/ZOOKEEPER-2628/src/java/main/org/apache/jute/compiler/JRecord.java#L530
Starting from these lines, compare both file side by side, they are the same. I am not sure why the diff was formatted this way.

hanm commented on src/java/main/org/apache/jute/compiler/JRecord.java:
```diff
@@ -403,168 +410,165 @@ public void genJavaCode(File outputDirectory) throws IOException {
                 JField jf = i.next();
-                jj.write(jf.genJavaCompareTo());
-                jj.write("    if (ret != 0) return ret;\n");
+                jj.write(jf.genJavaEquals());
```
Similarly, there is no actual change here.

hanm commented:
@fpj Thank you for taking time to double check the patch.

>> I noticed that there are also unaddressed comments from @lvfangmin and @eribeiro .
The comments were about using Interface type instead of implementation type, and I addressed the comments by creating ZOOKEEPER-2630 because this issue is not a regression and is not specifically about findbug warnings, and I think it might be better to address them separately.

I've left replies on your other comments in PR, please take a look.
